lockfile/src/sbom.rs

```diff
@@ -0,0 +1,222 @@
+#[serde(rename_all = "camelCase")]
+pub struct ExternalRefs {
+    pub reference_category: ReferenceCategory,
+    pub reference_locator: String,
+    pub reference_type: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, PartialOrd, Clone)]
+#[serde(rename_all = "SCREAMING-KEBAB-CASE")]
+pub enum ReferenceCategory {
+    Other,
+    // older schema uses _
+    #[serde(alias = "PERSISTENT_ID")]
+    PersistentId,
+    Security,
+    // older schema uses _
+    #[serde(alias = "PACKAGE_MANAGER")]
+    PackageManager,
+}
+
+impl TryFrom<&PackageInformation> for Package {
+    type Error = anyhow::Error;
+
+    fn try_from(pkg_info: &PackageInformation) -> anyhow::Result<Self> {
+        let pkg_url = pkg_info
+            .external_refs
+            .iter()
+            .find_map(|external| match external.reference_category {
+                ReferenceCategory::PackageManager => {
+                    if external.reference_type == "purl" {
+                        Some(&external.reference_locator)
+                    } else {
+                        None
+                    }
+                },
+                _ => None,
+            })
+            .context("Package manager not found")?;
+
+        let purl = PackageUrl::from_str(pkg_url).context("Unable to parse package url")?;
+        let package_type =
+            PackageType::from_str(purl.ty()).map_err(|_| anyhow!("Unrecognized ecosystem"))?;
+        let name = match purl.namespace() {
+            Some(ns) => format!("{}/{}", ns, purl.name()),
+            None => purl.name().into(),
+        };
+        let pkg_version = match (&pkg_info.version_info, purl.version()) {
+            (Some(v), _) => Some(v.to_string()),
+            (None, Some(v)) => Some(v.into()),
+            _ => None,
+        }
+        .context("Unable to determine version")?;
+
+        let version = purl
+            .qualifiers()
+            .iter()
+            .find_map(|(key, value)| match key.as_ref() {
+                "repository_url" => Some(PackageVersion::ThirdParty(ThirdPartyVersion {
+                    version: pkg_version.clone(),
+                    registry: value.to_string(),
+                })),
+                "download_url" => Some(PackageVersion::DownloadUrl(value.to_string())),
+                "vcs_url" => {
+                    if value.as_ref().starts_with("git+") {
+                        Some(PackageVersion::Git(value.to_string()))
+                    } else {
+                        None
+                    }
+                },
+                _ => None,
+            })
+            .unwrap_or(PackageVersion::FirstParty(pkg_version));
+
+        Ok(Package { name, version, package_type })
+    }
+}
+
+pub struct Sbom;
+
+impl Parse for Sbom {
+    fn parse(&self, data: &str) -> anyhow::Result<Vec<Package>> {
+        let mut lock: Spdx = serde_json::from_str(data).unwrap_or(serde_yaml::from_str(data)?);
```

Comment:
To be clear, this is equivalent to this:

```rust
let yaml_lock: Spdx = serde_yaml::from_str(data)?
let mut lock: Spdx = serde_json::from_str(data).unwrap_or(yaml_lock);
```

Which means that:

1. Any file which isn't valid YAML is immediately rejected
2. Every file is sent through both the `serde_yaml` and `serde_json` parser

This technically works because JSON is a subset of YAML. But I don't think this is what was desired.

I think you wanted one of the following options:

## Parse as JSON first
```suggestion
        let mut lock: Spdx = serde_json::from_str(data).or_else(|| serde_yaml::from_str(data))?;
```

## Always parse as YAML
```suggestion
        // This also parses JSON files because YAML is insane
        let mut lock: Spdx = serde_yaml::from_str(data)?;
```

Reply:
Ah yes, I meant the first option. Didn't realize it was being eagerly evaluated. Thanks!